app/domain/invoices/sac_memberships/extend_roles_for_invoicing.rb

```diff
@@ -15,28 +15,90 @@ class ExtendRolesForInvoicing
+        Role.where(id: batch.pluck(:id)).update_all(end_on: @prolongation_date)
       end
     end
 
     private
 
     def roles_to_extend
-      Role.with_inactive.where(type: ROLES_TO_EXTEND, terminated: false, end_on: ...@date, person_id: person_ids)
+      Role.where(type: ROLES_TO_EXTEND, terminated: false, end_on: ...@prolongation_date, person_id: person_ids)
+    end
+
+    def convert_roles_to_youth
+      roles_to_turn_youth.includes(:group, :person).find_each do |role|
+        if role.person.household.members.count == 2
+          role.person.household.people.each { leave_household(_1) }
+        else
+          leave_household(role.person)
+        end
+      end
+    end
+
+    def convert_roles_to_adult
+      roles_to_turn_adult.includes(:person, :group).find_each do |role|
+        end_role(role)
+        create_mitglied_role(role.person, role.group, :adult)
+
+        role.person.sac_membership.zusatzsektion_roles.where(terminated: false, end_on: ...@prolongation_date, beitragskategorie: :youth).find_each do |zusatzsektion_role|
+          end_role(zusatzsektion_role)
+          create_zusatzmitglied_role(zusatzsektion_role.person, zusatzsektion_role.group, :adult)
+        end
+      end
+    end
+
+    def leave_household(person)
+      Memberships::FamilyMutation.new(person, new_role_end_on: @prolongation_date, new_role_start_on: new_role_start_on, replaced_role_end_on: @reference_date).leave!
+    end
+
+    def roles_to_turn_youth
+      Role.joins(:person).where(type: SacCas::MITGLIED_STAMMSEKTION_ROLES.map(&:sti_name), terminated: false, end_on: ...@prolongation_date, beitragskategorie: :family, person_id: beitragskategorie_change_relevant_person_ids, person: {birthday: turned_youth_reference_age})
+    end
+
+    def roles_to_turn_adult
+      Role.joins(:person).where(type: SacCas::MITGLIED_STAMMSEKTION_ROLES.map(&:sti_name), terminated: false, end_on: ...@prolongation_date, beitragskategorie: :youth, person_id: beitragskategorie_change_relevant_person_ids, person: {birthday: turned_adult_reference_age})
+    end
+
+    def beitragskategorie_change_relevant_person_ids
+      Person.joins(:roles_unscoped)
+        .where(id: person_ids, roles: {end_on: ..@prolongation_date})
+        .where.not(roles: {start_on: @reference_date.., end_on: @prolongation_date..})
+        .select(:id)
     end
 
     def person_ids
       Person.joins(:roles_unscoped)
-        .where(roles: {type: Group::SektionsMitglieder::Mitglied.sti_name, terminated: false, end_on: ..@date})
-        .where.not(id: ExternalInvoice::SacMembership.where(year: @date.year).select(:person_id))
+        .where(roles: {type: Group::SektionsMitglieder::Mitglied.sti_name, terminated: false, end_on: ..@prolongation_date})
+        .where.not(id: ExternalInvoice::SacMembership.where(year: @prolongation_date.year).select(:person_id))
+        .where.not(Role.future.where(type: ROLES_TO_EXTEND).arel.exists)
```

Comment:
Ich bin unsicher betreffend dieser neuen Condition. Im roles_to_extend Fall braucht es sie imho nicht. Bei den neuen Fällen gibt es ja die beitragskategorie_change_relevant_person_ids.

Ausserdem fehlen Tests dazu. Beziehen sich z.B. diese future Roles überhaupt auf die Person oder wird nur geprüft, dass gar keine Future Roles in der gesamten DB sind?